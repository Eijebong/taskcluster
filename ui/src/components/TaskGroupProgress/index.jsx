import React, { Component } from 'react';
import { bool, func, shape, number, string } from 'prop-types';
import { sum } from 'ramda';
import { titleCase } from 'title-case';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import amber from '@material-ui/core/colors/amber';
import blue from '@material-ui/core/colors/blue';
import grey from '@material-ui/core/colors/grey';
import green from '@material-ui/core/colors/green';
import purple from '@material-ui/core/colors/purple';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import ButtonBase from '@material-ui/core/ButtonBase';
import CheckIcon from 'mdi-react/CheckIcon';
import ClockOutlineIcon from 'mdi-react/ClockOutlineIcon';
import CalendarClockIcon from 'mdi-react/CalendarClockIcon';
import AlertCircleOutlineIcon from 'mdi-react/AlertCircleOutlineIcon';
import CloseIcon from 'mdi-react/CloseIcon';
import AutorenewIcon from 'mdi-react/AutorenewIcon';
import PlaylistRemoveIcon from 'mdi-react/PlaylistRemoveIcon';
import Spinner from '../Spinner';
import { taskState } from '../../utils/prop-types';
import { TASK_STATE, THEME } from '../../utils/constants';
import Helmet from '../Helmet';

@withStyles(theme => ({
  statusButton: {
    display: 'flex',
    flexGrow: 1,
    flexBasis: 0,
    padding: `${theme.spacing(1)}px ${theme.spacing(1)}px`,
    justifyContent: 'space-around',
    cursor: 'pointer',
    margin: theme.spacing(1),
    '& > div': {
      display: 'flex',
      flexDirection: 'column',
      width: '100%',
    },
    borderRadius: theme.spacing(0.25),
  },
  statusButtonTypography: {
    color: THEME.PRIMARY_TEXT_DARK,
  },
  statusTitle: {
    textAlign: 'right',
  },
  statusIcon: {
    fill: THEME.PRIMARY_TEXT_DARK,
  },
  completedButton: {
    backgroundColor: theme.palette.success.dark,
    '&:hover': {
      backgroundColor: green[900],
    },
  },
  failedButton: {
    backgroundColor: theme.palette.error.main,
    '&:hover': {
      backgroundColor: theme.palette.error.dark,
    },
  },
  unscheduledButton: {
    backgroundColor: grey[600],
    '&:hover': {
      backgroundColor: grey[800],
    },
  },
  pendingButton: {
    backgroundColor: purple[400],
    '&:hover': {
      backgroundColor: purple[600],
    },
  },
  exceptionButton: {
    backgroundColor: theme.palette.warning.dark,
    '&:hover': {
      backgroundColor: amber[900],
    },
  },
  runningButton: {
    backgroundColor: blue[700],
    '&:hover': {
      backgroundColor: blue[900],
    },
  },
  noTasksButton: {
    opacity: 0.3,
  },
  spinner: {
    left: '47%',
    bottom: '42%',
    position: 'absolute',
  },
}))
export default class TaskGroupProgress extends Component {
  static propTypes = {
    /** The selected task state. This will change the card icon. */
    filter: taskState,
    /** Callback fired when the a state card is clicked */
    onStatusClick: func.isRequired,
    /**
     * If false, a spinner indicator will be displayed in the card.
     * This is useful to have when loading is done in steps
     * (e.g., 1000 tasks at a time)
     * */
    taskGroupLoaded: bool,
    /** The task group ID being inspected. */
    taskGroupId: string.isRequired,
    /** Status counts for each task state. */
    statusCount: shape({
      completed: number.isRequired,
      failed: number.isRequired,
      exception: number.isRequired,
      running: number.isRequired,
      pending: number.isRequired,
      unscheduled: number.isRequired,
    }).isRequired,
  };

  static defaultProps = {
    filter: null,
    taskGroupLoaded: true,
  };

  getStatusIcon = status => {
    if (this.props.filter === status) {
      return PlaylistRemoveIcon;
    }

    switch (status) {
      case TASK_STATE.COMPLETED: {
        return CheckIcon;
      }

      case TASK_STATE.EXCEPTION: {
        return AlertCircleOutlineIcon;
      }

      case TASK_STATE.RUNNING: {
        return AutorenewIcon;
      }

      case TASK_STATE.PENDING: {
        return ClockOutlineIcon;
      }

      case TASK_STATE.FAILED: {
        return CloseIcon;
      }

      case TASK_STATE.UNSCHEDULED: {
        return CalendarClockIcon;
      }

      default: {
        break;
      }
    }
  };

  getTaskGroupState = () => {
    const {
      completed,
      exception,
      failed,
      pending,
      running,
      unscheduled,
    } = this.props.statusCount;
    const allTasks = sum([completed, exception, pending, running, unscheduled]);
    const unfinishedTasks = sum([pending, running, unscheduled]);

    if (allTasks === 0) {
      return;
    }

    if (failed > 0 || exception > 0) {
      return TASK_STATE.FAILED;
    }

    if (unfinishedTasks > 0) {
      return TASK_STATE.RUNNING;
    }

    return TASK_STATE.COMPLETED;
  };

  render() {
    const { classes, onStatusClick, taskGroupLoaded, statusCount } = this.props;
    const showDots = Object.values(statusCount).reduce((a, b) => a + b) === 0;
    const taskGroupState = this.getTaskGroupState();

    return (
      <Grid container spacing={2}>
        <Helmet state={taskGroupState} />
        {Object.keys(TASK_STATE).map(status => {
          const Icon = this.getStatusIcon(status);
          const count = statusCount[status.toLowerCase()];

          return (
            <ButtonBase
              focusRipple
              key={status}
              name={status}
              variant="contained"
              onClick={onStatusClick}
              className={classNames(
                classes[`${status.toLowerCase()}Button`],
                classes.statusButton,
                {
                  [classes.noTasksButton]: count === 0,
                }
              )}>
              <div>
                <Icon color="white" className={classes.statusIcon} size={32} />
                {!taskGroupLoaded && !showDots && (
                  <Spinner size={12} className={classes.spinner} />
                )}
              </div>
              <div>
                <Typography
                  align="right"
                  className={classes.statusButtonTypography}
                  variant="h4">
                  {showDots ? '...' : count}
                </Typography>
                <Typography
                  className={classNames(
                    classes.statusTitle,
                    classes.statusButtonTypography
                  )}
                  variant="caption">
                  {titleCase(status)}
                </Typography>
              </div>
            </ButtonBase>
          );
        })}
      </Grid>
    );
  }
}
