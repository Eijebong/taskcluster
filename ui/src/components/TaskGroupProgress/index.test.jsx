import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import TaskGroupProgress from './index';

it('should render TaskGroupProgress', () => {
  const nop = () => {};
  const { asFragment } = render(
    <MemoryRouter keyLength={0}>
      <TaskGroupProgress
        taskGroupId="abc"
        taskGroupLoaded={false}
        taskGroup={{ edges: [] }}
        filter="RUNNING"
        onStatusClick={nop}
        onUpdate={nop}
      />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
