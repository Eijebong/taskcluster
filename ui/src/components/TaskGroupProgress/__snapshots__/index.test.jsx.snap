// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TaskGroupProgress 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <button
      class="MuiButtonBase-root TaskGroupProgress-completedButton-5 TaskGroupProgress-statusButton-1 TaskGroupProgress-noTasksButton-11"
      name="COMPLETED"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <div>
        <svg
          class="mdi-icon TaskGroupProgress-statusIcon-4"
          fill="white"
          height="32"
          viewBox="0 0 24 24"
          width="32"
        >
          <path
            d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"
          />
        </svg>
      </div>
      <div>
        <h4
          class="MuiTypography-root TaskGroupProgress-statusButtonTypography-2 MuiTypography-h4 MuiTypography-alignRight"
        >
          ...
        </h4>
        <span
          class="MuiTypography-root TaskGroupProgress-statusTitle-3 TaskGroupProgress-statusButtonTypography-2 MuiTypography-caption"
        >
          COMPLETED
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root TaskGroupProgress-runningButton-10 TaskGroupProgress-statusButton-1 TaskGroupProgress-noTasksButton-11"
      name="RUNNING"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <div>
        <svg
          class="mdi-icon TaskGroupProgress-statusIcon-4"
          fill="white"
          height="32"
          viewBox="0 0 24 24"
          width="32"
        >
          <path
            d="M2,6V8H14V6H2M2,10V12H11V10H2M14.17,10.76L12.76,12.17L15.59,15L12.76,17.83L14.17,19.24L17,16.41L19.83,19.24L21.24,17.83L18.41,15L21.24,12.17L19.83,10.76L17,13.59L14.17,10.76M2,14V16H11V14H2Z"
          />
        </svg>
      </div>
      <div>
        <h4
          class="MuiTypography-root TaskGroupProgress-statusButtonTypography-2 MuiTypography-h4 MuiTypography-alignRight"
        >
          ...
        </h4>
        <span
          class="MuiTypography-root TaskGroupProgress-statusTitle-3 TaskGroupProgress-statusButtonTypography-2 MuiTypography-caption"
        >
          RUNNING
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root TaskGroupProgress-failedButton-6 TaskGroupProgress-statusButton-1 TaskGroupProgress-noTasksButton-11"
      name="FAILED"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <div>
        <svg
          class="mdi-icon TaskGroupProgress-statusIcon-4"
          fill="white"
          height="32"
          viewBox="0 0 24 24"
          width="32"
        >
          <path
            d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
          />
        </svg>
      </div>
      <div>
        <h4
          class="MuiTypography-root TaskGroupProgress-statusButtonTypography-2 MuiTypography-h4 MuiTypography-alignRight"
        >
          ...
        </h4>
        <span
          class="MuiTypography-root TaskGroupProgress-statusTitle-3 TaskGroupProgress-statusButtonTypography-2 MuiTypography-caption"
        >
          FAILED
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root TaskGroupProgress-pendingButton-8 TaskGroupProgress-statusButton-1 TaskGroupProgress-noTasksButton-11"
      name="PENDING"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <div>
        <svg
          class="mdi-icon TaskGroupProgress-statusIcon-4"
          fill="white"
          height="32"
          viewBox="0 0 24 24"
          width="32"
        >
          <path
            d="M12,20A8,8 0 0,0 20,12A8,8 0 0,0 12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22C6.47,22 2,17.5 2,12A10,10 0 0,1 12,2M12.5,7V12.25L17,14.92L16.25,16.15L11,13V7H12.5Z"
          />
        </svg>
      </div>
      <div>
        <h4
          class="MuiTypography-root TaskGroupProgress-statusButtonTypography-2 MuiTypography-h4 MuiTypography-alignRight"
        >
          ...
        </h4>
        <span
          class="MuiTypography-root TaskGroupProgress-statusTitle-3 TaskGroupProgress-statusButtonTypography-2 MuiTypography-caption"
        >
          PENDING
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root TaskGroupProgress-exceptionButton-9 TaskGroupProgress-statusButton-1 TaskGroupProgress-noTasksButton-11"
      name="EXCEPTION"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <div>
        <svg
          class="mdi-icon TaskGroupProgress-statusIcon-4"
          fill="white"
          height="32"
          viewBox="0 0 24 24"
          width="32"
        >
          <path
            d="M11,15H13V17H11V15M11,7H13V13H11V7M12,2C6.47,2 2,6.5 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20Z"
          />
        </svg>
      </div>
      <div>
        <h4
          class="MuiTypography-root TaskGroupProgress-statusButtonTypography-2 MuiTypography-h4 MuiTypography-alignRight"
        >
          ...
        </h4>
        <span
          class="MuiTypography-root TaskGroupProgress-statusTitle-3 TaskGroupProgress-statusButtonTypography-2 MuiTypography-caption"
        >
          EXCEPTION
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root TaskGroupProgress-unscheduledButton-7 TaskGroupProgress-statusButton-1 TaskGroupProgress-noTasksButton-11"
      name="UNSCHEDULED"
      tabindex="0"
      type="button"
      variant="contained"
    >
      <div>
        <svg
          class="mdi-icon TaskGroupProgress-statusIcon-4"
          fill="white"
          height="32"
          viewBox="0 0 24 24"
          width="32"
        >
          <path
            d="M15,13H16.5V15.82L18.94,17.23L18.19,18.53L15,16.69V13M19,8H5V19H9.67C9.24,18.09 9,17.07 9,16A7,7 0 0,1 16,9C17.07,9 18.09,9.24 19,9.67V8M5,21C3.89,21 3,20.1 3,19V5C3,3.89 3.89,3 5,3H6V1H8V3H16V1H18V3H19A2,2 0 0,1 21,5V11.1C22.24,12.36 23,14.09 23,16A7,7 0 0,1 16,23C14.09,23 12.36,22.24 11.1,21H5M16,11.15A4.85,4.85 0 0,0 11.15,16C11.15,18.68 13.32,20.85 16,20.85A4.85,4.85 0 0,0 20.85,16C20.85,13.32 18.68,11.15 16,11.15Z"
          />
        </svg>
      </div>
      <div>
        <h4
          class="MuiTypography-root TaskGroupProgress-statusButtonTypography-2 MuiTypography-h4 MuiTypography-alignRight"
        >
          ...
        </h4>
        <span
          class="MuiTypography-root TaskGroupProgress-statusTitle-3 TaskGroupProgress-statusButtonTypography-2 MuiTypography-caption"
        >
          UNSCHEDULED
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
